mygrad/nnet/layers/recurrent.py

```diff
@@ -158,3 +159,170 @@ def simple_RNN(X, U, W, s0=None, bp_lim=None):
+        else:
+            new_dat = np.vstack((self._z.data, z[1:]))
+            self._z = Tensor(new_dat, _creator=self)
+
+        if not self._r:
+            self._r = Tensor(r, _creator=self)
+        else:
+            new_dat = np.vstack((self._r.data, r[1:]))
+            self._r = Tensor(new_dat, _creator=self)
+
+        if not self._h:
+            self._h = Tensor(h, _creator=self)
+        else:
+            new_dat = np.vstack((self._h.data, h[1:]))
+            self._h = Tensor(new_dat, _creator=self)
+
+        return self._hidden_seq
+
+
+    def backward(self, grad, seq_index=None):
+        s = self._hidden_seq.data[:-1]
+        z = self._z.data
+        r = self._r.data
+        h = self._h.data
+
+        dLds = grad[1:]
+        old_dLds = np.zeros_like(dLds)
+
+        """
+        for i in range(self.bp_lim):
+            # dL_{n} / ds_{t+1} -> dL_{n} / df_{t+1}  | ( n > t )
+            index = slice(2, len(grad) - i)
+            dLn_ft1 = dst_dft[index] * (dLt_dst[index] - old_dst[index])
+            old_dst = np.copy(dLt_dst)
+            dLt_dst[1:len(grad) - (i + 1)] += np.dot(dLn_ft1, self.W.data.T)  # dL_{t} / ds_{t} + ... + dL_{n} / ds_{t}
+        """
+
+        for i in range(min(s.shape[0] - 1, self.bp_lim)):
+            #  dL(t) / ds(t) + dL(t+1) / ds(t)
+            dt = dLds[1:len(dLds) - i] - old_dLds[1:len(dLds) - i]
+            old_dLds = np.copy(dLds)
+            dLds[:len(dLds) - (i + 1)] += _gru_dLds(s[1:len(dLds) - i],
+                                                   z[1:len(dLds) - i],
+                                                   r[1:len(dLds) - i],
+                                                   h[1:len(dLds) - i],
+                                                   dt,
+                                                   self.Wz.data,
+                                                   self.Wr.data,
+                                                   self.Wh.data)
+
+        zgrad = dLds * (s - h) # dL / dz
+        hgrad = dLds * (1 - z) # dL / dh
+        rgrad = np.dot((1 - h ** 2) * hgrad, self.Wh.data.T) * s # dL / dr
+
+        self._hidden_seq.grad = dLds
+        self._z.grad = zgrad
+        self._r.grad = rgrad
+        self._h.grad = hgrad
+
+        dz = zgrad * z * (1 - z)
+        dr = rgrad * r * (1 - r)
+        dh = hgrad * (1 - h ** 2)
+
+        self.Uz.backward(np.einsum("ijk, ijl -> kl", self._input_seq, dz))
+        self.Wz.backward(np.einsum("ijk, ijl -> kl", s, dz))
+        self.bz.backward(np.sum(np.sum(dz, axis=0), axis=0))
+
+        self.Ur.backward(np.einsum("ijk, ijl -> kl", self._input_seq, dr))
+        self.Wr.backward(np.einsum("ijk, ijl -> kl", s, dr))
+        self.br.backward(np.sum(np.sum(dr, axis=0), axis=0))
+
+        self.Uh.backward(np.einsum("ijk, ijl -> kl", self._input_seq, dh))
+        self.Wh.backward(np.einsum("ijk, ijl -> kl", (s * r), dh))
+        self.bh.backward(np.sum(np.sum(dh, axis=0), axis=0))
```

Comment:
I just realized that we need to backprop through `X` too! In case you chain GRUs together (i.e. the hidden descriptors of the first GRU is `X` for the subsequent)